File: Cistem.java
import java.util.regex.Pattern;
public class Cistem {;

	private static final Pattern GE_PATTERN = Pattern.compile("^ge(.{4,})");
	private static final Pattern DOLLAR1_PATTERN = Pattern.compile("(.)\\1");
	private static final Pattern ND_PATTERN = Pattern.compile("nd$");
	private static final Pattern EMR_PATTERN = Pattern.compile("e[mr]$");
	private static final Pattern T_PATTERN = Pattern.compile("t$");
	private static final Pattern ESN_PATTERN = Pattern.compile("[esn]$");
	private static final Pattern STAR_PATTERN = Pattern.compile("(.)\\*");

	public static String stem(String word) {
		return stem(word, false);
	}

	public static String stem(String word, boolean case_insensitive) {
		if (word.length() == 0) return word;

		word = word.replace("Ü", "U");
		word = word.replace("Ö", "O");
		word = word.replace("Ä", "A");
		word = word.replace("ü", "u");
		word = word.replace("ö", "o");
		word = word.replace("ä", "a");

		boolean uppercase = Character.isUpperCase(word.charAt(0));

		word = word.toLowerCase();

		word = word.replace("ß", "ss");
		word = GE_PATTERN.matcher(word).replaceAll("$1");
		word = word.replace("sch", "$");
		word = word.replace("ei", "%");
		word = word.replace("ie", "&");

		word = DOLLAR1_PATTERN.matcher(word).replaceAll("$1*");

		while (word.length() > 3) {
			if (word.length() > 5) {
				String newWord = EMR_PATTERN.matcher(word).replaceAll("");
				if (!word.equals(newWord)) {
					word = newWord;
					continue;
				}

				newWord = ND_PATTERN.matcher(word).replaceAll("");
				if (!word.equals(newWord)) {
					word = newWord;
					continue;
				}
			}

			if (!uppercase || case_insensitive) {
				final String newWord = T_PATTERN.matcher(word).replaceAll("");
				if (!word.equals(newWord)) {
					word = newWord;
					continue;
				}
			}

			String newWord = ESN_PATTERN.matcher(word).replaceAll("");
			if (!word.equals(newWord)) {
				word = newWord;
				continue;
			} else {
				break;
			}
		}

		word = STAR_PATTERN.matcher(word).replaceAll("$1$1");
		word = word.replace("&", "ie");
		word = word.replace("%", "ei");
		word = word.replace("$", "sch");

		return word;
	}

	public static String[] segment(String word) {
		return segment(word, false);
	}

	public static String[] segment(String word, boolean case_insensitive) {
		if (word.length() == 0) {
			String[] result = new String[2];
			result[0] = "";
			result[1] = "";
			return result;
		}

		int restLength = 0;
		boolean uppercase = Character.isUpperCase(word.charAt(0));
		word = word.toLowerCase();
		String original = new String(word);

		word = word.replace("sch", "$");
		word = word.replace("ei", "%");
		word = word.replace("ie", "&");

		word = DOLLAR1_PATTERN.matcher(word).replaceAll("$1*");

		while (word.length() > 3) {
			if (word.length() > 5) {
				String newWord = word.replaceAll("e[mr]$", "");
				if (!word.equals(newWord)) {
					restLength += 2;
					word = newWord;
					continue;
				}

				newWord = word.replaceAll("nd$", "");
				if (!word.equals(newWord)) {
					restLength += 2;
					word = newWord;
					continue;
				}
			}

			if (!uppercase || case_insensitive) {
				String newWord = word.replaceAll("t$", "");
				if (!word.equals(newWord)) {
					restLength += 1;
					word = newWord;
					continue;
				}
			}

			String newWord = word.replaceAll("[esn]$", "");
			if (!word.equals(newWord)) {
				restLength += 1;
				word = newWord;
				continue;
			} else {
				break;
			}
		}

		word = word.replaceAll("(.)\\*", "$1$1");
		word = word.replace("&", "ie");
		word = word.replace("%", "ei");
		word = word.replace("$", "sch");

		String rest = "";
		if (restLength != 0) {
			rest = original.substring(original.length() - restLength);
		}

		String[] result = new String[2];
		result[0] = word;
		result[1] = rest;
		return result;
	}
}

File: Cistem.cs
using System;
using System.Text.RegularExpressions;

namespace Cistem
{
  public static class Cistem
  {
    private const string GE_PATTERN = "^ge(.{4,})";
    private const string DOLLAR1_PATTERN = "(.)\\1";
    private const string ND_PATTERN = "nd$";
    private const string EMR_PATTERN = "e[mr]$";
    private const string T_PATTERN = "t$";
    private const string ESN_PATTERN = "[esn]$";
    private const string STAR_PATTERN = "(.)\\*";

    public static String Stem(String word)
    {
      return Stem(word, false);
    }

    public static String Stem(String word, bool case_insensitive)
    {
      if (word.Length == 0) return word;

      word = word.Replace("Ü", "U");
      word = word.Replace("Ö", "O");
      word = word.Replace("Ä", "A");
      word = word.Replace("ü", "u");
      word = word.Replace("ö", "o");
      word = word.Replace("ä", "a");

      var uppercase = char.IsUpper(word[0]);

      word = word.ToLower();

      word = word.Replace("ß", "ss");
      word = Regex.Replace(word, GE_PATTERN, "$1");
      word = word.Replace("sch", "$");
      word = word.Replace("ei", "%");
      word = word.Replace("ie", "&");

      word = Regex.Replace(word, DOLLAR1_PATTERN, "$1*");

      string newWord;

      while (word.Length > 3)
      {
        if (word.Length > 5)
        {
          newWord = Regex.Replace(word, EMR_PATTERN, "");
          if (word != newWord)
          {
            word = newWord;
            continue;
          }

          newWord = Regex.Replace(word, ND_PATTERN, "");
          if (word != newWord)
          {
            word = newWord;
            continue;
          }
        }

        if (!uppercase || case_insensitive)
        {
          newWord = Regex.Replace(word, T_PATTERN, "");
          if (word != newWord)
          {
            word = newWord;
            continue;
          }
        }

        newWord = Regex.Replace(word, ESN_PATTERN, "");

        if (word != newWord)
          word = newWord;
        else
          break;
      }

      word = Regex.Replace(word, STAR_PATTERN, "$1$1");
      word = word.Replace("&", "ie");
      word = word.Replace("%", "ei");
      word = word.Replace("$", "sch");

      return word;
    }

    public static String[] Segment(String word)
    {
      return Segment(word, false);
    }

    public static String[] Segment(String word, bool case_insensitive)
    {
      if (word.Length == 0)
        return new string[] { string.Empty, string.Empty };

      var restLength = 0;
      var uppercase = char.IsUpper(word[0]);
      word = word.ToLower();
      var original = word;

      word = word.Replace("sch", "$");
      word = word.Replace("ei", "%");
      word = word.Replace("ie", "&");

      word = Regex.Replace(word, DOLLAR1_PATTERN, "$1*");

      string newWord;

      while (word.Length > 3)
      {
        if (word.Length > 5)
        {
          newWord = Regex.Replace(word, EMR_PATTERN, "");
          if (word != newWord)
          {
            restLength += 2;
            word = newWord;
            continue;
          }

          newWord = Regex.Replace(word, ND_PATTERN, "");
          if (word != newWord)
          {
            restLength += 2;
            word = newWord;
            continue;
          }
        }

        if (!uppercase || case_insensitive)
        {
          newWord = Regex.Replace(word, T_PATTERN, "");
          if (word != newWord)
          {
            restLength += 1;
            word = newWord;
            continue;
          }
        }

        newWord = Regex.Replace(word, ESN_PATTERN, "");
        if (word != newWord)
        {
          restLength += 1;
          word = newWord;
        }
        else
          break;
      }

      word = Regex.Replace(word, STAR_PATTERN, "$1$1");
      word = word.Replace("&", "ie");
      word = word.Replace("%", "ei");
      word = word.Replace("$", "sch");

      var rest = string.Empty;

      if (restLength != 0)
        rest = original.Substring(original.Length - restLength);

      return new string[] { word, rest };
    }
  }
}